### PR TITLE
Refactor vast::type and vast::data

### PR DESCRIPTION
**Pull request by mavam**
This commit is a followup of #193:
- [x] Port `vast::type`
- [x] Port `vast::data`
- [x] Remove `vast::variant`

### CODE REVIEW

**Comment by Neverlord** [comment]
@mavam I could not convince myself that `polymorphic_visit` is something we would want to have long-term in the code base. I've implemented the `caf::sum_type_access<vast::type>` trait instead to treat `type` as regular sum type. This gives us the old-and-trusted `visit` back and I've reverted as much code as possible.

Dispatching is now done via an `index()`, just like we do in variants. This allows us simply dispatch once through a function-pointer table (see `sum_type_access::dispatcher`) and no longer requires RTTI. I'd say we drop the `polymorphic_visitor` from the code base again (it is unused now). What do you think?

**Comment by mavam** [comment]
I agree that the previous form of dispatching is a bliss to work with in comparison to the new polymorphic visitation API. While it sucks that I sunk so much time in this, I'm actually glad to see that you already starting reverting the changes. The previous API simply made much more sense, double dispatch was a bliss, and it worked great with type deduction and templates. In other words, it was productive. The polymorphic visitation API feels like it's in only the way.

Long story short, I'm onboard. 👍
